platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,440 @@
+
+  private static final RetryPolicy RETRY_UNTIL_NOT_INTERRUPTED_AND_NO_ERROR_AND_NOT_NULL =
+      new RetryPolicy()
+          .retryWhen(null)
+          .retryOn(SolrServerException.class, SolrException.class, IOException.class)
+          .abortOn(InterruptedIOException.class, InterruptedException.class)
+          .withBackoff(10L, TimeUnit.MINUTES.toMillis(1L), TimeUnit.MILLISECONDS);
+
+  private static final long ERROR_MIN_FREQUENCY = TimeUnit.MINUTES.toMillis(1L);
+
+  private static final UnavailableSolrClient CLOSED_CLIENT =
+      new UnavailableSolrClient("Solr client was closed");
+
+  private static final ScheduledExecutorService EXECUTOR = SolrClientAdapter.createExecutor();
+
+  private final Object lock = new Object();
+
+  private final String core;
+
+  private final Creator creator;
+
+  private final List<Listener> listeners = new CopyOnWriteArrayList<>();
+
+  private final Queue<Initializer> initializers = new ConcurrentLinkedQueue<>();
+
+  // writes and reads are always protected by synchronization
+  @Nullable private Future<SolrClient> future = null;
+
+  // writes are always protected by synchronization, reads are not
+  private volatile SolrClientProxy client;
+
+  private final AtomicLong lastError = new AtomicLong();
+
+  /**
+   * Constructs a new client adapter for the specified code and using the specified creator to
+   * create new Solr client instances.
+   *
+   * <p><i>Note:</i> There is no need to implement any retry behavior in the creator as this will be
+   * handled by this class. Simply attempt the creation and fail fast. The creation of the client
+   * can be interrupted (see {@link Thread#interrupted}) in which case it should attempt to stop the
+   * creation as quickly as possible. It is acceptable for the creator to throw back any of the
+   * following exceptions: {@link IOException}, {@link SolrServerException}, {@link
+   * org.apache.solr.common.SolrException}, {@link InterruptedException}, or {@link
+   * InterruptedIOException}. A retry will automatically be triggered if returning <code>null</code>
+   * or any exceptions other than {@link InterruptedException} or {@link InterruptedIOException} or
+   * thrown back.
+   *
+   * @param core the Solr core for which to create an adaptor
+   * @param creator the creator to use for creating corresponding Solr clients
+   * @throws IllegalArgumentException if <code>core</code>, <code>info</code>, and <code>creator
+   *     </code> is <code>null</code>
+   */
+  public SolrClientAdapter(String core, Creator creator) {
+    Validate.notNull(core, "invalid null core");
+    Validate.notNull(creator, "invalid null creator");
+    LOGGER.debug("Solr({}): Creating a Solr client adapter", core);
+    this.core = core;
+    this.creator = creator;
+    try {
+      setClient(new UnavailableSolrClient("initializing '" + core + "' core"), true);
+    } catch (IOException e) { // should never happen as exceptions are swallowed above
```

Comment:
:question: Even though we think this will never happen, should we put a log just in case code is changed and it does happen?